enhancements/builds/volume-secrets.md

```diff
@@ -0,0 +1,225 @@
+
+There are a few undesirable aspects to this:
+1) In the dockerfile case, the content can still be found in lower layers of the image unless a layer squashing option is 
+selected.
+2) Requires extra work by the user in the Dockerfile, so each Dockerfile must be customized
+
+This enhancement proposes to introduce an option to use buildah's capability to mount a volume at build time.  The content mounted into the build pod would be then mounted into the container processing the Dockerfile, making that content available within the container so Dockerfile commands could reference it.  No explicit `ADD` would be required, and since mounted content is not committed to the resulting image, no `RUN rm` equivalent is required to clean up the injected content.
+
+
+## Motivation
+
+### Goals
+
+* Simplify how users consume secret + configmap content in builds
+* Increase the security of protected content being injected to images
+* Simplify use cases like mounting subscription entitlement credentials into a build so subscription content can be pulled down during a build while not including the credentials in the resulting image.
+* Eventually extend this api to allow the mounting of traditional volumes (such as those backed by persistent storage)
+
+### Non-Goals
+
+* This enhancement should not result in a change of behavior for users of the existing secret/configmap injection api.
+
+
+## Proposal
+
+### User Stories [optional]
+
+The enabled use cases are essentially identical to what can be done with the configmap/secret input api in builds today, but with a better user experience and security as discussed above.  It does not enable a new use case that is not already possible today, except that layer squashing will not be required.
+
+Future extensions to this enhancement could enable the use case of providing build input content from a persistent volume and allowing the build to store/cache content for future builds on such a volume.  Those will be discussed in the future enhancement at that time.
+
+### Implementation Details/Notes/Constraints [optional]
+
+We will need to introduce a new mechanism in the build api which allows the user to indicate that they want to inject "volume" content into the build.  Initially the only allowed volume types will be configmaps and secrets.  The api will otherwise be similar to the existing secret/configmap injection api in which users identify the configmap/secret and the target path for injection.  
+
+This will be done by adding a Volume[] field to the Source and Docker strategy structs.  The Volume[] field will allow 
+defining volumes to be mounted to the build pod in the same way that a normal pod allows for this.  Similarly a VolumeMount[]
+field will be added, but without the MountPropagation and SubPath fields.  MountPropagation and SubPath can be considered
+for support in the future.
```

Comment:
I wonder if there will be both some API reuse as well as perhaps some level of pod template manipulation with this that could be shared with build v2 if it was factored in the right way (both api and repo housing code)

How interested are you @bparees in citing that as a goal above?

Reply:
i'd say "not very".  If we decide we got this api perfect, then i'm ok w/ builds v2 just cloning/copying it as it's just not that complicated/bulky.  I'd rather not try to abstract it in hopes of reusing it, especially in case we ultimately conclude it's "close" but not perfect by the time we get to the builds v2 api.

In any case the only new structs this is actually defining is the VolumeMount struct (simplified version of the k8s VolumeMount).  So that actually will be a standalone struct that could in theory be reused elsewhere if desired.

Reply:
Reading over volumes support in Tekton, what they provide appears to be a thin vaneer over the kube volume API. We need to be clear that the mount path is the destination _as seen inside the build_, not the path that we mount into the container that runs buildah.

Reply:
> We need to be clear that the mount path is the destination as seen inside the build, not the path that we mount into the container that runs buildah.

yes good point.  the build-controller will autogen/pick the pod mount path.  the VolumeMount will determine the mapping of those pod mount paths, to the buildah args.

Reply:
i was trying to make that clear with `  // Path within the build environment at which the volume should be mounted.`